### PR TITLE
module implementation

### PR DESCRIPTION
**Pull request by thiesen**

### CODE REVIEW

File: README.tfdoc.hcl [hunk 1]
```diff
@@ -93,145 +90,217 @@ section {
...
+            description = <<-END
+                List of metadata fields that should be added to reported logs. Can only be specified if VPC flow logs for this subnetwork is `enabled` and `"metadata"` is set to `CUSTOM_METADATA`.
+              END
+          }
+
+          attribute "filter_expr" {
+            type        = string
+            description = <<-END
+                Export filter used to define which VPC flow logs should be logged, as as CEL expression. See https://cloud.google.com/vpc/docs/flow-logs#filtering for details on how to format this field.
+              END
```
Comment (kassianh) [inline README.tfdoc.hcl:301]:
```suggestion
            END
```

File: README.tfdoc.hcl [hunk 2]
```diff
@@ -93,145 +90,217 @@ section {
...
+            description = <<-END
+                Can only be specified if VPC flow logging for this subnetwork is `enabled`. Configures whether metadata fields should be added to the reported VPC flow logs. Possible values are `EXCLUDE_ALL_METADATA`, `INCLUDE_ALL_METADATA`, and `CUSTOM_METADATA`.
+              END
+          }
+
+          attribute "metadata_fields" {
+            type        = list(string)
+            description = <<-END
+                List of metadata fields that should be added to reported logs. Can only be specified if VPC flow logs for this subnetwork is `enabled` and `"metadata"` is set to `CUSTOM_METADATA`.
+              END
```
Comment (kassianh) [inline README.tfdoc.hcl:294]:
```suggestion
            END
```

File: README.tfdoc.hcl [hunk 3]
```diff
@@ -93,145 +90,217 @@ section {
...
+            description = <<-END
+                Can only be specified if VPC flow logging for this subnetwork is enabled. The value of the field must be in `[0, 1]`. Set the sampling rate of VPC flow logs within the subnetwork where `1.0` means all collected logs are reported and `0.0` means no logs are reported.
+              END
+          }
+
+          attribute "metadata" {
+            type        = string
+            description = <<-END
+                Can only be specified if VPC flow logging for this subnetwork is `enabled`. Configures whether metadata fields should be added to the reported VPC flow logs. Possible values are `EXCLUDE_ALL_METADATA`, `INCLUDE_ALL_METADATA`, and `CUSTOM_METADATA`.
+              END
```
Comment (kassianh) [inline README.tfdoc.hcl:287]:
```suggestion
            END
```

File: README.tfdoc.hcl [hunk 4]
```diff
@@ -93,145 +90,217 @@ section {
...
+            description = <<-END
+                Can only be specified if VPC flow logging for this subnetwork is enabled. Toggles the aggregation interval for collecting flow logs. Increasing the interval time will reduce the amount of generated flow logs for long lasting connections. Default is an interval of `5 seconds` per connection. Possible values are `INTERVAL_5_SEC`, `INTERVAL_30_SEC`, `INTERVAL_1_MIN`, `INTERVAL_5_MIN`, `INTERVAL_10_MIN`, and `INTERVAL_15_MIN`.
+              END
+          }
+
+          attribute "flow_sampling" {
+            type        = number
+            description = <<-END
+                Can only be specified if VPC flow logging for this subnetwork is enabled. The value of the field must be in `[0, 1]`. Set the sampling rate of VPC flow logs within the subnetwork where `1.0` means all collected logs are reported and `0.0` means no logs are reported.
+              END
```
Comment (kassianh) [inline README.tfdoc.hcl:280]:
```suggestion
            END
```

File: README.tfdoc.hcl [hunk 5]
```diff
@@ -93,145 +90,217 @@ section {
...
+                metadata_fields      = "CUSTOM_METADATA"
+                filter_expr          = true
+              }
+            END
+
+          attribute "aggregation_interval" {
+            type        = string
+            description = <<-END
+                Can only be specified if VPC flow logging for this subnetwork is enabled. Toggles the aggregation interval for collecting flow logs. Increasing the interval time will reduce the amount of generated flow logs for long lasting connections. Default is an interval of `5 seconds` per connection. Possible values are `INTERVAL_5_SEC`, `INTERVAL_30_SEC`, `INTERVAL_1_MIN`, `INTERVAL_5_MIN`, `INTERVAL_10_MIN`, and `INTERVAL_15_MIN`.
+              END
```
Comment (kassianh) [inline README.tfdoc.hcl:273]:
```suggestion
            END
```

File: README.tfdoc.hcl [hunk 6]
```diff
@@ -93,145 +90,217 @@ section {
...
+
+          readme_example = <<-END
+              log_config {
+                aggregation_interval = "INTERVAL_10_MIN"
+                flow_sampling        = 0.5
+                metadata             = "INCLUDE_ALL_METADATA"
+                metadata_fields      = "CUSTOM_METADATA"
+                filter_expr          = true
+              }
+            END
```
Comment (kassianh) [inline README.tfdoc.hcl:267]:
```suggestion
          END
```

File: README.tfdoc.hcl [hunk 7]
```diff
@@ -93,145 +90,217 @@ section {
...
+                The name associated with this subnetwork secondary range, used when adding an alias IP range to a VM instance. The name must be 1-63 characters long, and comply with RFC1035. The name must be unique within the subnetwork.
+              END
+          }
+
+          attribute "ip_cidr_range" {
+            required    = true
+            type        = string
+            description = <<-END
+                The range of IP addresses belonging to this subnetwork secondary range. Provide this property when you create the subnetwork. Ranges must be unique and non-overlapping with all primary and secondary IP ranges within a network. Only `IPv4` is supported.
+              END
```
Comment (kassianh) [inline README.tfdoc.hcl:249]:
```suggestion
            END
```

File: README.tfdoc.hcl [hunk 8]
```diff
@@ -93,145 +90,217 @@ section {
...
+                ip_cidr_range = "192.168.10.0/24"
+              }
+            END
+
+          attribute "range_name" {
+            required    = true
+            type        = string
+            description = <<-END
+                The name associated with this subnetwork secondary range, used when adding an alias IP range to a VM instance. The name must be 1-63 characters long, and comply with RFC1035. The name must be unique within the subnetwork.
+              END
```
Comment (kassianh) [inline README.tfdoc.hcl:241]:
```suggestion
            END
```

File: README.tfdoc.hcl [hunk 9]
```diff
@@ -93,145 +90,217 @@ section {
...
+          type           = list(secondary_ip_range)
+          description    = <<-END
+              An array of configurations for secondary IP ranges for VM instances contained in this subnetwork. The primary IP of such VM must belong to the primary ipCidrRange of the subnetwork. The alias IPs may belong to either primary or secondary ranges.
+            END
+          readme_example = <<-END
+              secondary_ip_range {
+                range_name    = "tf-test-secondary-range-update1"
+                ip_cidr_range = "192.168.10.0/24"
+              }
+            END
```
Comment (kassianh) [inline README.tfdoc.hcl:234]:
```suggestion
          END
```

File: README.tfdoc.hcl [hunk 10]
```diff
@@ -93,145 +90,217 @@ section {
...
-        readme_example = <<-END
-          module_depends_on = [
-            null_resource.name
-          ]
-        END
+        variable "secondary_ip_ranges" {
+          type           = list(secondary_ip_range)
+          description    = <<-END
+              An array of configurations for secondary IP ranges for VM instances contained in this subnetwork. The primary IP of such VM must belong to the primary ipCidrRange of the subnetwork. The alias IPs may belong to either primary or secondary ranges.
+            END
```
Comment (kassianh) [inline README.tfdoc.hcl:228]:
```suggestion
          END
```

File: README.tfdoc.hcl [hunk 11]
```diff
@@ -93,145 +90,217 @@ section {
...
 
-        attribute "delete" {
+        variable "ip_cidr_range" {
+          required    = true
           type        = string
           description = <<-END
-            Timeout for delete operations.
-          END
+              The range of internal addresses that are owned by this subnetwork. Provide this property when you create the subnetwork. For example, 10.0.0.0/8 or 192.168.0.0/16. Ranges must be unique and non-overlapping within a network. Only IPv4 is supported.
+            END
```
Comment (kassianh) [inline README.tfdoc.hcl:221]:
```suggestion
          END
```

File: README.tfdoc.hcl [hunk 12]
```diff
@@ -93,145 +90,217 @@ section {
...
+              The GCP region for this subnetwork.
+            END
+        }
+
+        variable "private_ip_google_access" {
+          type        = bool
+          default     = true
+          description = <<-END
+              When enabled, VMs in this subnetwork without external IP addresses can access Google APIs and services by using Private Google Access.
+            END
```
Comment (kassianh) [inline README.tfdoc.hcl:213]:
```suggestion
          END
```

File: README.tfdoc.hcl [hunk 13]
```diff
@@ -93,145 +90,217 @@ section {
...
 
-        attribute "update" {
+        variable "region" {
+          required    = true
           type        = string
           description = <<-END
-            Timeout for update operations.
-          END
+              The GCP region for this subnetwork.
+            END
```
Comment (kassianh) [inline README.tfdoc.hcl:205]:
```suggestion
          END
```

File: README.tfdoc.hcl [hunk 14]
```diff
@@ -93,145 +90,217 @@ section {
...
+        }
 
-        attribute "create" {
+        variable "description" {
           type        = string
           description = <<-END
-            Timeout for create operations.
-          END
+              An optional description of this subnetwork. Provide this property when you create the resource. This field can be set only at resource creation time.
+            END
```
Comment (kassianh) [inline README.tfdoc.hcl:197]:
```suggestion
          END
```

File: README.tfdoc.hcl [hunk 15]
```diff
@@ -93,145 +90,217 @@ section {
...
-          module_timeouts = {
-            null_resource = {
-              create = "4m"
-              update = "4m"
-              delete = "4m"
-            }
-          }
-        END
+            Specifically, the name must be 1-63 characters long and match the regular expression `[a-z]([-a-z0-9]*[a-z0-9])?` which means the first character must be a lowercase letter, and all following characters must be a dash, lowercase letter, or digit, except the last character, which cannot be a dash.
+            END
```
Comment (kassianh) [inline README.tfdoc.hcl:190]:
```suggestion
          END
```

File: README.tfdoc.hcl [hunk 16]
```diff
@@ -93,145 +90,217 @@ section {
     END
 
     section {
-      title = "Main Resource Configuration"
+      title = "Top-level Arguments"
+
+      section {
+        title = "Module Configuration"
```
Comment (kassianh) [inline README.tfdoc.hcl:96]: I believe this section should go last the series of these sub-sections.

File: README.tfdoc.hcl [hunk 17]
```diff
@@ -93,145 +90,217 @@ section {
     END
 
     section {
-      title = "Main Resource Configuration"
+      title = "Top-level Arguments"
```
Comment (kassianh) [inline README.tfdoc.hcl:93]: According to the template this should be two indented sections:

`Module Argument Reference`
- `Main Resource Configuration`

File: README.tfdoc.hcl [hunk 18]
```diff
@@ -63,25 +50,35 @@ section {
...
     title   = "Module Features"
     content = <<-END
-      This module implements the following Terraform resources:
-
-      - `null_resource`
-
-      and supports additional features of the following modules:
-
-      - [mineiros-io/something/google](https://github.com/mineiros-io/terraform-google-something)
+      A [Terraform] module for creating `google_compute_subnetwork` resources
```
Comment (kassianh) [inline README.tfdoc.hcl:53]: Maybe we could add the list of resources that the module implements, not sure how much it matters to follow the template, it's just what I've been doing when creating new documentation.

File: Makefile [hunk 19]
```diff
@@ -105,7 +105,7 @@ test/unit-tests:
 ## Generate README.md with Terradoc
 .PHONY: terradoc
 terradoc:
-	$(call quiet-command,terradoc -o README.md README.tfdoc.hcl)
+	$(call quiet-command,terradoc generate -o README.md README.tfdoc.hcl)
```
Comment (kassianh) [inline Makefile:108]: @mariux I've made myself a Shortcut task to update this in the other repo's.

File: README.tfdoc.hcl [hunk 20]
```diff
@@ -293,7 +358,7 @@ section {
     END
 
     section {
-      title   = "Backwards compatibility in `0.0.z` and `0.y.z` version"
+      title   = "Backwards compatibility in ` 0.0.z ` and ` 0.y.z ` version"
```
Comment (kassianh) [inline README.tfdoc.hcl:361]:
```suggestion
      title   = "Backwards compatibility in `0.0.z` and `0.y.z` version"
```

**Review by kassianh — CHANGES_REQUESTED** [review]
A few suggestions for the documentation. The actual module implementation looks good to me :)

File: README.tfdoc.hcl (hunk 18 above)
Comment (thiesen) [reply README.tfdoc.hcl]: you're right! this module was based on https://github.com/mineiros-io/terraform-google-network-subnet so the template used to create it was outdated; I updated it accordingly! thanks for the catch, man!

File: Makefile (hunk 19 above)
Comment (soerenmartius) [reply Makefile:108]: Mabe checkout https://github.com/gruntwork-io/git-xargs for automating such tasks across several repositories :)

Comment (kassianh) [reply Makefile:108]: @soerenmartius Perfect thank you! Was gonna ask if there was a tool to make this easier

**Review by kassianh — DISMISSED** [review]

File: README.tfdoc.hcl [hunk 21]
```diff
@@ -95,63 +90,143 @@ section {
...
+            The ID of the project in which the resources belong. If it is not set, the provider project is used.
+          END
+      }
 
-      # TODO: remove examples
+      variable "network" {
+        required    = true
+        type        = string
+        description = <<-END
+            The VPC network the subnets belong to. Only networks that are in the distributed mode can have subnetworks.
```
Comment (soerenmartius) [inline README.tfdoc.hcl:104]: Can we make this more clear? What's required? The `id` or the `name` of the network?

**Review by soerenmartius — CHANGES_REQUESTED** [review]

File: README.tfdoc.hcl [hunk 22]
```diff
@@ -95,63 +90,143 @@ section {
...
+              The GCP region for this subnetwork.
+            END
+      }
+
+      variable "private_ip_google_access" {
+        type        = bool
+        default     = true
+        description = <<-END
+              When enabled, VMs in this subnetwork without external IP addresses can access Google APIs and services by using Private Google Access.
+            END
```
[the remark this answers is not in the document]
Comment (thiesen) [reply README.tfdoc.hcl]: the [terraform-google-network-subnetwork module](https://github.com/mineiros-io/terraform-google-network-subnet) was initially used as the base for this module and I made the changes in that version; I only realized that the doc was outdated after @kassianh review and updated it according with the [latest template](https://github.com/mineiros-io/terraform-module-template) but I missed some problems like the one you pointed here :sweat_smile:

**Review by soerenmartius — APPROVED** [review]
LGTM! Good job @thiesen

File: versions.tf [hunk 23]
```diff
@@ -1,14 +1,14 @@
-# ----------------------------------------------------------------------------------------------------------------------
+# ---------------------------------------------------------------------------------------------------------------------
 # SET TERRAFORM AND PROVIDER REQUIREMENTS FOR RUNNING THIS MODULE
-# ----------------------------------------------------------------------------------------------------------------------
+# ---------------------------------------------------------------------------------------------------------------------
 
 terraform {
-  required_version = "~> 1.0"
+  required_version = ">= 0.14, < 2.0"
```
Comment (soerenmartius) [inline versions.tf:6]: I think we can leave this to 
```
  required_version = "~> 1.0"
```

as we don't need to support Terraform <1.0 anymore

**Review by soerenmartius — APPROVED** [review]
lgtm! good job

File: main.tf [hunk 24]
```diff
@@ -1,5 +1,51 @@
...
+
+  dynamic "log_config" {
+    for_each = try(var.log_config, {})
+
+    content {
+      aggregation_interval = try(each.value.aggregation_interval, null)
+      flow_sampling        = try(each.value.flow_sampling, null)
+      metadata             = try(each.value.metadata, null)
+      metadata_fields      = try(each.value.metadata_fields, null)
+      filter_expr          = try(each.value.filter_expr, null)
```
Comment (mariux) [inline main.tf:36]: `each.value` is not available here

File: main.tf [hunk 25]
```diff
@@ -1,5 +1,51 @@
...
+resource "google_compute_subnetwork" "subnetwork" {
+  for_each = var.module_enabled ? 1 : 0
+
+  project     = var.project
+  network     = var.network
+  region      = var.region
+  name        = var.name
+  description = try(var.description, null)
+
+  private_ip_google_access = try(var.private_ip_google_access, true)
```
Comment (mariux) [inline main.tf:16]: try can not be used on variables..

**Pull request merged**